File: hands-on/src/mat_mul.c
#include <stdio.h> // needed for ‘printf’ 
#include <stdlib.h> // needed for ‘RAND_MAX’ 
#include <omp.h> // needed for OpenMP 
#include <time.h> // needed for clock() and CLOCKS_PER_SEC etc
#include "helper.h" // local helper header to clean up code


void initialize_matrices(float** A, float** B, float** C, int ROWS, int COLUMNS){
    for (int i = 0 ; i < ROWS ; i++)
    {
        for (int j = 0 ; j < COLUMNS ; j++)
        {
            A[i][j] = (float) rand() / RAND_MAX ;
            B[i][j] = (float) rand() / RAND_MAX ;
            C[i][j] = 0.0 ;
        }
    }
}

void simple_matrix_multiply(float** A, float** B, float** C, int ROWS, int COLUMNS){
    
    printf("(Simple) Matix Multiplication of 2D matricies of equal sizes (%d, %d)\n",ROWS,COLUMNS);

    for(int i=0;i<ROWS;i++)
    {
        for(int j=0;j<COLUMNS;j++)
        {
            for(int k=0;k<COLUMNS;k++)
            {
                C[i][j] += A[i][k]*B[k][j];
            }
        }
    }
}

void openmp_matrix_multiply(float** A, float** B, float** C, int ROWS, int COLUMNS){
    
    int num_threads = omp_get_max_threads();
    
    printf("(OpenMP) Matix Multiplication of 2D matricies of equal sizes (%d, %d)\n",ROWS,COLUMNS);
    printf("Using %d Threads\n", num_threads);
    
    #pragma omp parallel for 
    for (int i = 0; i < ROWS; ++i) 
    {
        for (int j = 0; j < COLUMNS; ++j) 
        {
            for (int k = 0; k < COLUMNS; ++k) 
            {
                C[i][j] = C[i][j] + A[i][k] * B[k][j];
            }
        }
    }
}


int main( int argc, char *argv[] )  {

  int ROWS;
  int COLUMNS;
  int N;

  /* DUMB bools needed for the argument parsing logic */
  bool openmp = false;
  bool simple = true;
  bool sanity_check = false;
  
  /* VERY DUMB Argument Parsers */
  N = parse_arguments(argc, argv, &simple, &openmp, &sanity_check);
  ROWS = N;
  COLUMNS = N;
  /* declare the arrays */
  float **A;
  float **B;
  float **C;

  /* allocate the arrays */
  A = malloc(ROWS * sizeof *A);
  B = malloc(ROWS * sizeof *B);
  C = malloc(ROWS * sizeof *C);
  for (int i=0; i<ROWS; i++)
  {
    A[i] = malloc(COLUMNS * sizeof *A[i]);
    B[i] = malloc(COLUMNS * sizeof *B[i]);
    C[i] = malloc(COLUMNS * sizeof *C[i]);
  }

  /*======================================================================*/
  /*                START of Section of the code that matters!!!          */
  /*======================================================================*/

  /* initialize the arrays */
  initialize_matrices(A, B, C, ROWS, COLUMNS);

  /* Simple matrix multiplication */
  /*==============================*/
  if (true == simple)
  {
    clock_t t; // declare clock_t (long type)
    t = clock(); // start the clock
    
    simple_matrix_multiply(A, B, C, ROWS, COLUMNS);
    
    t = clock() - t; // stop the clock

    double time_taken = ((double)t)/CLOCKS_PER_SEC; // convert to seconds (and long to double)
    printf("TIME: %f sec\n",time_taken);
  }



  /* OpenMP parallel matrix multiplication */
  /*=======================================*/
  if (true == openmp)
  {
    // omp_get_wtime needed here because clock will sum up time for all threads
    double start = omp_get_wtime();  

    openmp_matrix_multiply(A, B, C, ROWS, COLUMNS);
    
    double end = omp_get_wtime(); 
    printf("TIME: %f sec\n",(end-start));
  }

  /*======================================================================*/
  /*                 END of Section of the code that matters!!!           */
  /*======================================================================*/

  /* deallocate the arrays */
  for (int i=0; i<ROWS; i++)
  {
    free(A[i]);
    free(B[i]);
    free(C[i]);
  }
  free(A);
  free(B);
  free(C);
}

File: hands-on/bin/mat_mul.py
# Simple Matrix multiplication 
import sys
import numpy as np
import time
import argparse
from random import random


def simple_matrix_multiply(A,B):
    """
    Very Naive implementation of Matrix Multiplication.

    :Matrix A: list of lists A 
    :Matrix B: list of lists B 
    :return: Matrix C
    """ 

    C = [[0 for x in range(len(A))] for y in range(len(B[0]))]
    for i in range(len(A)):
        for j in range(len(B[0])):
            for k in range(len(B)):
                C[i][j] += A[i][k] * B[k][j]
    return C

def numpy_matrix_multiply(A,B):
    """
    Simple NumPy implementation of Matrix Multiplication.

    :Matrix A: list of lists A 
    :Matrix B: list of lists B 
    :return: Matrix C
    """ 

    npA = np.array(A)
    npB = np.array(B)
    C = np.matmul(A,B)
    return C

if __name__ == "__main__":

    # Parse the arguments
    parser = argparse.ArgumentParser()
    parser.add_argument("ROWS", help="N rows",type=int)
    parser.add_argument("COLUMNS", help="N Columns",type=int)
    parser.add_argument("-s", "--simple", help="Use Simple Matmul (Default if no option given)",action="store_true",default=True)
    parser.add_argument("-n", "--numpy", help="Use NumPy Matmul ",action="store_true")
    args = parser.parse_args()
    
    #Set matrix dimensions
    ROWS = args.ROWS
    COLUMNS = args.COLUMNS

    #Define Matrix A ((list of lists))
    A = [[random() for x in range(ROWS)] for y in range(COLUMNS)]
    #Define Matrix B (list of lists)
    B = [[random() for x in range(ROWS)] for y in range(COLUMNS)]

    # choose simple mat mul
    if args.simple and not args.numpy: #Dumb logic here
        print("(Simple) Matrix Multiplication of 2D matricies of equal sizes ("+str(ROWS)+", "+str(COLUMNS)+")")

        #  start the clock
        start = time.perf_counter_ns()

        simple_matrix_multiply(A,B)
        
        #  stop the clock
        end = time.perf_counter_ns()
        print("Time of Serial is " +str(round(end-start)/1e9) + " s") # convert to seconds

    # choose NumPy mat mul
    if args.numpy:
        print("(NumPy) Matrix Multiplication of 2D matricies of equal sizes ("+str(ROWS)+", "+str(COLUMNS)+")")

        #  start the clock
        start = time.perf_counter_ns()
        
        numpy_matrix_multiply(A,B)

        #  stop the clock
        end = time.perf_counter_ns()
        print("Time of Numpy is " +str(round(end-start)/1e9) + " s") # convert to seconds

